### DISCUSSION TITLE
From cron_expr struct to const char *

### DISCUSSION THREAD

**Discussion by Lepiqpiq** [exander77/supertinycron#1] [Ideas]
Hello,

First of all, I would like to say great job for this lib.
What do you think of creating a function that take a cron_expr structure as a paramater and return a string corresponding to the cron_expr structure that you pass in the function ? 
const char * CronExprToString(struct cron_expr * cron);
I think that would be helpfull in certain application.

Thank you,

**Comment by exander77** [comment]
@Lepiqpiq What is the case? You want to make `cron_expr` by hand and see the expression that can produce it?
I see two issues:
1. Many `cron_expr` structures will have many expressions that can produce them.
2. Many `cron_expr` structures will have no expressions that can produce them.

**Reply by exander77** [reply]
I can see as a way to test the library thoroughly by making conversion back and forth.

Btw, I pushed massive rewrite of the library with parser written from scratch (loosely written LL1 parser), merged more prev/next code, removed dynamic memory allocation etc. Code shrank by several hundreds of lines.

**Reply by Lepiqpiq** [reply]
Thank you for you response.
I see what you mean. The idea is that I'm storing a cron_expr and I want to send it as a json. So I need this cron_expr as a string. 

I'll have a look to your massive commit.
THank you

**Reply by Lepiqpiq** [reply]
Concerning the issue of having many possibilities to write cron_expr, I was thinking of taking a paramater that specifies the string format you want to have. Like, a flag to write range instead of values, write days instead of week days...

**Reply by exander77** [reply]
@Lepiqpiq I was thinking about it a little as well, I can use such feature for extensive testing making it convert back and forth. There could be flags if you would like to use values, or ranges, use ordinals or numbers as well.

For your use case, why don't use store strings representation as well? Or We can support storing original expression within cron_expr.

```
typedef struct {
    uint8_t seconds[8];
    uint8_t minutes[8];
    uint8_t hours[3];
    uint8_t days_of_week[1];
    uint8_t days_of_month[4];
    int8_t  day_in_month[1];
    /**     
     * Flags:       
     * 0 last day of the month
     * 1 last weekday of the month
     * 2 closest weekday to day in month
     */     
    uint8_t flags[1];
    uint8_t years[EXPR_YEARS_LENGTH];
    uint8_t months[2];
    chart* expression;
} cron_expr;
```

It would cost almost nothing to add an expression field that would basically pass the input back for easy access.

**Reply by exander77** [reply]
I was also thinking that if I were to reconstruct the original expression I may as well produce textual representation like:
"every monday at midnight".

**Reply by Lepiqpiq** [reply]
Using store strings means storing const char. I don't have much memory that's why I don't want to store string and want to generate a temp string from a struct (~20bytes) instead of 100-200 bytes if I store the string.

It's good idea to generate a function for verbosity :)

**Reply by exander77** [reply]
@Lepiqpiq This reminds me, I added years support which takes additional 29 bytes, I may introduce a macro for low memory footprint to disable this feature. If you do not have much memory, like embedded hardware, those 29 bytes for year may be a pain. Something like `CRON_DISABLE_YEARS`.

**Reply by exander77** [reply]
@Lepiqpiq So I added `CRON_DISABLE_YEARS`, if that is turned on, it will disable years functionality. It will work with 7 fields, but years are ignored (meaning `*`).
Without years, it takes 30 bytes, with years it takes 59 bytes.

**Reply by Lepiqpiq** [reply]
For years that's a good idea. I know that some application doesn't take into account second also.
Thank you for you help. Let met know if you include a feature that take a cron_expr and that convert it into a cron string.

**Reply by exander77** [reply]
@Lepiqpiq It seems viable to implement in some limited form.

**Reply by exander77** [reply]
@Lepiqpiq I did initial implementation of `cron_generate_expr`. And doesn't check anything and doesn't produce any error. Caller is responsible to provide buffer of appropriate size and nothing is checked, so it can segfault on memory overflow. Will improve in the future. You can check it for now. It also doesn't have any arguments and provides single type of output.

**Reply by exander77** [reply]
@Lepiqpiq Now it accepts maximum length and reports failure.

**Reply by Lepiqpiq** [reply]
@exander77 Thank you ! I'll try this. I let you know if I'm facing some unhandled errors or whatever that might help to improve the lib :)

**Reply by Lepiqpiq** [reply]
@exander77 cron_generate_expr does not support option "CRON_DISABLE_YEARS"
Just a forgotten #ifndef CRON_DISABLE_YEARS line 805

**Reply by exander77** [reply]
@Lepiqpiq You are correct, I pushed a fix.

**Reply by exander77** [reply]
@Lepiqpiq I have to add tests that run compilation with CRON_DISABLE_YEARS to be sure I don't break it in the future.

**Reply by exander77** [reply]
@Lepiqpiq Btw, what do you use the library for?

**Reply by exander77** [reply]
@Lepiqpiq It should be pretty stable, you can see calls in the `ccronepr_test.c`.
I run all the test expressions through parse, generate, parse and then compare the structure produces. All test cases pass.

**Comment by exander77** [comment]
@Lepiqpiq I improved this functionality, so it has some support for repeats as well.
